### PR TITLE
Added SshCommand.InputStream to allow writing to stdin of SshCommand

### PR DESCRIPTION
**Pull request by realvizu**
Changes:
- Added SshCommand.InputStream (type: ChannelInputStream), which sends the data written to the stream through a command's channel.
- Added an integration test that echos back the sent data using the "cat" command.
- All unit tests pass.

I you have recommendations for further unit tests, please advise.

---

I started the work by trying to run all unit tests on my dev machine to see if they all pass. Unfortunately a lot of them failed, so I had to make the following modifications (see the changes in my first commit in this PR):
- Excluded MD5 tests for the net462 target, because I got System.InvalidOperationException: 'This implementation is not part of the Windows Platform FIPS validated cryptographic algorithms.'
- Modified SshdConfig.cs: do not throw for "Include", just do nothing.
- Modified failing dos2unix parameters in Dockerfile.TestServer.
- Added a line to .gitattributes to force LF line ending for key files used by integration tests, otherwise using them causes error.
- Modified SftpClientTest.Test_Sftp_Multiple_Async_Upload_And_Download_10Files_5MB_Each to upload only 2 files instead of 10 because for more than 2 files I got timeout errors.

If you know a better fix for any of the above modifications, please advise.

### CODE REVIEW

File: test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs
Comment (WojciechNagorski) [inline test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs]: I'll have to check this on my machine because it seems to be working now. But I know that this test is unstable. so don't worry about it.

File: test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs
Comment (WojciechNagorski) [inline test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs]: I'll have to check this on my machine because it seems to be working now.

File: test/Renci.SshNet.Tests/Classes/PrivateKeyFileTest.cs
Comment (WojciechNagorski) [inline test/Renci.SshNet.Tests/Classes/PrivateKeyFileTest.cs]: I'll have to check this on my machine because it seems to be working now.

File: test/Renci.SshNet.IntegrationTests/Dockerfile.TestServer
Comment (WojciechNagorski) [inline test/Renci.SshNet.IntegrationTests/Dockerfile.TestServer:14]: Maybe there is a new alpine image and it has different behavior, but I don't have a problem with this change as long as it works.

**Comment by WojciechNagorski** [comment]
It looks good to me! But I need more time to test it!

Do you have a specific function in mind that you want to implement using this functionality? 
Did you see that it can be useful in https://github.com/dotnet/Docker.DotNet/issues/540

**Comment by WojciechNagorski** [comment]
image
Those tests work on my machine.

File: src/Renci.SshNet/Common/ChannelInputStream.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/ChannelInputStream.cs]: This does not look right. What if `Write` is called multiple times? Can another test be added which writes one byte at a time to validate that? It is a valid scenario when calling `stream.CopyTo` which calls `Write` under the hood

File: test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs
Comment (Rob-Hague) [reply test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs]: It also fails on my machine, but I've not gotten to the bottom of why (I think sshd is crashing but maybe SSH.NET is doing something wrong). I think we should revert this change for this PR

File: test/Renci.SshNet.TestTools.OpenSSH/SshdConfig.cs
Comment (Rob-Hague) [inline test/Renci.SshNet.TestTools.OpenSSH/SshdConfig.cs]: No need for the comment

```suggestion
                case "Include":
```

File: test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs
Comment (Rob-Hague) [reply test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs]: Seems like @realvizu has FIPS enabled. We should probably condition these tests for when [CryptoConfig.AllowOnlyFipsAlgorithms](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.cryptoconfig.allowonlyfipsalgorithms) is false. In the meantime, can we revert these changes for this PR?

File: test/Renci.SshNet.Tests/Renci.SshNet.Tests.csproj
Comment (Rob-Hague) [inline test/Renci.SshNet.Tests/Renci.SshNet.Tests.csproj]: If we revert the FIPS test changes, then this is already unnecessary, but FYI the constant `NET462` is defined [by default](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives#conditional-compilation) so one can do `#if NET462` without any additional configuration

**Review by Rob-Hague — COMMENTED** [review]
I don't think we should send channel EOF in `Write`. How about calling it only in `Dispose` and initialising `InputStream` lazily, something like:

```c#
var cmd = ...
using (cmd.InputStream) // This initialises InputStream
{
    myStream.CopyTo(cmd.InputStream);
} // This disposes InputStream and sends EOF

```

?

File: test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs
Comment (WojciechNagorski) [reply test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs]: Yes, I agree because this is not a part of this PR. @realvizu Can you revert this?

File: test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs
Comment (WojciechNagorski) [reply test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs]: Yes, I agree because this is not a part of this PR. @realvizu Can you revert this?

File: test/Renci.SshNet.Tests/Classes/PrivateKeyFileTest.cs
Comment (WojciechNagorski) [reply test/Renci.SshNet.Tests/Classes/PrivateKeyFileTest.cs]: @realvizu Can you revert this? It is not a part of this PR.

File: test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs
Comment (realvizu) [reply test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs]: Sure, reverted it.

File: test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs
Comment (realvizu) [reply test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs]: Sure, reverted it.

File: test/Renci.SshNet.Tests/Classes/PrivateKeyFileTest.cs
Comment (realvizu) [reply test/Renci.SshNet.Tests/Classes/PrivateKeyFileTest.cs]: Sure, reverted it.

**Comment by realvizu** [comment]
> It looks good to me! But I need more time to test it!
> 
> Do you have a specific function in mind that you want to implement using this functionality? Did you see that it can be useful in [dotnet/Docker.DotNet#540](https://github.com/dotnet/Docker.DotNet/issues/540)

I had a very specific use case: I had to call some legacy system that is implemented as a set of SSH commands and one of the commands requires sending a file via input stream. These changes made it work.

File: src/Renci.SshNet/Common/ChannelInputStream.cs
Comment (realvizu) [reply src/Renci.SshNet/Common/ChannelInputStream.cs]: You're absoloutely right, I moved the EOF sending to Dispose, and added a unit test that writes the input one byte at a time.

File: test/Renci.SshNet.TestTools.OpenSSH/SshdConfig.cs
Comment (realvizu) [reply test/Renci.SshNet.TestTools.OpenSSH/SshdConfig.cs]: OK, removed the comment.

File: test/Renci.SshNet.Tests/Renci.SshNet.Tests.csproj
Comment (realvizu) [reply test/Renci.SshNet.Tests/Renci.SshNet.Tests.csproj]: OK, removed it.

**Comment by realvizu** [comment]
Thanks a lot @WojciechNagorski and @Rob-Hague for the comments and suggestions!

- I reverted all the FIPS-related changes and the change made to SftpClientTest.Upload.cs.
- I moved the EOF sending to Dispose and added a unit test that validates that now multiple calls to Write works as expected.
- I've adapted @Rob-Hague 's suggestion about making InputStream lazily initialized with some modification. I wanted to make it more apparent that a resource is created here (that should be disposed after using), so I changed the InputStream from a property to a CreateInputStream method.
- Also, because the input stream uses the channel created in BeginExecute and closed in EndExecute, I added a check that throws invalid operation exception if CreateInputStream is called when there's no channel to use. Also added some unit tests for this.

File: src/Renci.SshNet/Common/ChannelInputStream.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/ChannelInputStream.cs]: Since the constructor is internal, perhaps the type could just be internal unless there is a reason to make it public

File: src/Renci.SshNet/Common/ChannelInputStream.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/ChannelInputStream.cs]: Maybe remove this remark :-)

File: src/Renci.SshNet/Common/ChannelInputStream.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/ChannelInputStream.cs]: This should guard on `disposing`

```suggestion
                if (disposing && _totalPosition > 0 && _channel?.IsOpen == true)
```

File: src/Renci.SshNet/SshCommand.cs
Comment (Rob-Hague) [inline src/Renci.SshNet/SshCommand.cs]: Perhaps it should throw an error if it's already not-null, instead of silently doing nothing?

**Review by Rob-Hague — APPROVED** [review]
Nice

File: src/Renci.SshNet/Common/ChannelInputStream.cs
Comment (realvizu) [reply src/Renci.SshNet/Common/ChannelInputStream.cs]: Sure.

File: src/Renci.SshNet/Common/ChannelInputStream.cs
Comment (realvizu) [reply src/Renci.SshNet/Common/ChannelInputStream.cs]: Sure.

File: src/Renci.SshNet/Common/ChannelInputStream.cs
Comment (realvizu) [reply src/Renci.SshNet/Common/ChannelInputStream.cs]: Right.

File: src/Renci.SshNet/SshCommand.cs
Comment (realvizu) [reply src/Renci.SshNet/SshCommand.cs]: Right.

**Comment by realvizu** [comment]
Shall I do a squash+rebase onto develop before merging the PR?

**Review by Rob-Hague — APPROVED** [review]

**Comment by Rob-Hague** [comment]
It will be merged with "Squash and merge" so it is not necessary

(it requires a CODEOWNERS approval before it can merge)

**Review by WojciechNagorski — APPROVED** [review]

**Comment by WojciechNagorski** [comment]
There is a build problem:
> C:\projects\ssh-net\test\Renci.SshNet.IntegrationTests\SshClientTests.cs(37,33): error CS7036: There is no argument given that corresponds to the required parameter 'offset' of 'Stream.Write(byte[], int, int)' [C:\projects\ssh-net\test\Renci.SshNet.IntegrationTests\Renci.SshNet.IntegrationTests.csproj::TargetFramework=net48]

File: test/Renci.SshNet.IntegrationTests/SshClientTests.cs
Comment (Rob-Hague) [inline test/Renci.SshNet.IntegrationTests/SshClientTests.cs]:
```suggestion
                    inputStream.Write(inputByteArray, 0, inputByteArray.Length);
```

Comment (WojciechNagorski) [reply test/Renci.SshNet.IntegrationTests/SshClientTests.cs]: Thanks

**Comment by WojciechNagorski** [comment]
This issue has been fixed in the 2024.0.0 version.

**Pull request merged**
